### ISSUE TITLE
Harnessing LSTM for Stock Price Prediction: A Decade of Reliance Data

### ISSUE THREAD

**Issue by Agrawalvishesh68** [Agrawalvishesh68/Algorithimic-Trading-Strategies#1]
Predicting stock prices has long fascinated economists and data scientists. Markets are noisy and nonlinear, driven by short-term swings and long-term trends. Capturing these dynamics requires models that can retain context over time, something Long Short-Term Memory (LSTM) networks handle exceptionally well.

Why LSTM? 📊 

Standard recurrent neural networks process sequences but often forget distant information due to vanishing or exploding gradients. LSTMs solve this with memory cells and gates that decide what to keep, forget, or output.

-> Sigmoid (0–1): Powers input, forget, and output gates. Values near 0 block information; near 1 allow it.

-> Tanh (–1 to 1): Shapes the cell state, encoding positive or negative relationships and balancing updates.

Together, these activations let LSTMs filter noise and preserve long-term dependencies, making them ideal for financial time-series forecasting.

Project Overview

Objective: Forecast Reliance Industries’ daily stock prices using 10 years of data.

Data Prep: Downloaded historical prices from Yahoo Finance, scaled values with MinMaxScaler, and created 60-day input windows to predict the next day’s close.

Model: Keras Sequential network with two stacked LSTM layers, dropout for regularization, and the Adam optimizer with mean-squared-error loss. Initial 3-epoch runs validated setup, followed by 50 epochs for final training.

Evaluation: Root Mean Squared Error (RMSE) measured accuracy; predicted vs. actual prices were visually compared.

Key Results: RMSE dropped from ~69 after 10 epochs to ~37 by epoch 50, showing steady learning.

-> Outperformed moving-average baselines, capturing nuanced market patterns.

-> Remained stable under hyperparameter tweaks.

-> Predicted prices closely tracked actual trends.

Takeaways

LSTM’s memory advantage helps capture market signals that unfold over weeks, but quality preprocessing is critical. While powerful, the model is no crystal ball, sudden policy shifts or global events can defy historical patterns. Used wisely, such forecasts provide guidance, not guarantees, supporting investment strategy when paired with sound analysis.

<img width="632" height="412" alt="Image" src="https://github.com/user-attachments/assets/e82c1ee6-93eb-46cc-b3d7-c0d3ad0715ee" />

[LSTM Price Prediction model.ipynb](https://github.com/user-attachments/files/22441402/LSTM.Price.Prediction.model.ipynb)
